src/decode/ykpt/packet_parser/mod.rs

```diff
@@ -0,0 +1,303 @@
+            Self::Normal => &[
+                PacketKind::ShortTNT,
+                PacketKind::PAD,
+                PacketKind::FUP,
+                PacketKind::TIP,
+                PacketKind::CYC,
+                PacketKind::LongTNT,
+                PacketKind::PSB,
+                PacketKind::MODE,
+                PacketKind::TIPPGE,
+                PacketKind::TIPPGD,
+            ],
+            Self::PSBPlus => &[PacketKind::CBR, PacketKind::PSBEND],
+        }
+    }
+
+    /// Check if we need to transition state after parsing a given packet kind.
+    fn transition(&mut self, pkt_kind: PacketKind) {
+        let new = match (*self, pkt_kind) {
+            (Self::Init, PacketKind::PSB) => Self::PSBPlus,
+            (Self::Normal, PacketKind::PSB) => Self::PSBPlus,
+            (Self::PSBPlus, PacketKind::PSBEND) => Self::Normal,
+            _ => return, // No state transition.
+        };
+        *self = new;
+    }
+}
+
+pub(super) struct PacketParser<'t> {
+    /// The raw bytes of the PT trace we are iterating over.
+    bytes: &'t [u8],
+    /// The parser operates as a state machine. This field keeps track of which state we are in.
+    state: PacketParserState,
+    /// The most recent Target IP (TIP) value that we've seen. This is needed because updated TIP
+    /// values are sometimes compressed using bits from the previous TIP value.
+    prev_tip: usize,
+}
+
+/// Attempt to read the packet of type `$packet` using deku. On success wrap the packet up into the
+/// corresponding discriminant of `Packet`.
+macro_rules! read_to_packet {
+    ($packet: ty, $bits: expr, $discr: expr) => {
+        <$packet>::read($bits, ()).and_then(|(r, p)| Ok((r, $discr(p))))
+    };
+}
+
+/// Same as `read_to_packet!`, but with extra logic for dealing with packets which encode a TIP.
+macro_rules! read_to_packet_tip {
+    ($packet: ty, $bits: expr, $discr: expr, $prev_tip: expr) => {
+        <$packet>::read($bits, ()).and_then(|(r, p)| {
+            let ret = if p.needs_prev_tip() {
+                Ok((r, $discr(p, Some($prev_tip))))
+            } else {
+                Ok((r, $discr(p, None)))
+            };
+            ret
+        })
+    };
+}
+
+impl<'t> PacketParser<'t> {
+    pub(super) fn new(bytes: &'t [u8]) -> Self {
+        Self {
+            bytes,
+            state: PacketParserState::Init,
+            prev_tip: 0,
+        }
+    }
+
+    fn mk_parse_error(&mut self) -> HWTracerError {
+        let mut s = String::new();
+        write!(
+            &mut s,
+            "In state '{:?}', failed to parse packet: ",
+            self.state
+        )
+        .unwrap();
+        write!(&mut s, "{}", self.byte_stream_str(8, ", ")).unwrap();
+        HWTracerError::TraceParseError(s)
+    }
+
+    /// Attempt to parse a packet of the specified `PacketKind`.
+    fn parse_kind(&mut self, kind: PacketKind) -> Option<Packet> {
+        let bits = BitSlice::from_slice(self.bytes).ok()?;
```

Comment:
`ok` looks suspicious here, as it's normally used to mean "ignore errors".

Reply:
`ok` converts from `Result` to `Option`.

I've used it here because I don't see value in propagating the error from the bitvec crate up, when `None` is sufficient for our purposes.

Reply:
My point is: why would we get an `Err` and why are we ignoring it? Ignoring it without a comment explaining why raises so many questions!

Reply:
I don't see it as ignoring it. We return the exceptional value `None` to indicate the error, but I it's not interesting to know why it failed.

FWIW:
https://docs.rs/bitvec/0.22.3/bitvec/slice/struct.BitSlice.html#returns-2

If you really want to propagate it, we can, but we'd have to either make our own composite error type (compounding BitVec and deku error types), or add those error types to `HWTracerError`.

Note that we do the same below where we discard a deku error and return None.

Reply:
Ah, I see. OK with me!